peps/pep-0793.rst

```diff
@@ -0,0 +1,476 @@
+(We expect functions to export a static constant, or one of several constants
+chosen depending on, for example, ``Py_Version``. Dynamic behaviour should
+generally happen in the ``Py_mod_create`` and ``Py_mod_exec`` functions.)
+
+
+Dynamic creation
+----------------
+
+A new function will be added to create a module from an array of slots:
+
+.. code-block:: c
+
+    PyObject *PyModule_FromSlotsAndSpec(PyModuleDef_Slot *slots, PyObject *spec)
+
+The *slots* argument must point to an array of ``PyModuleDef_Slot`` structures,
+terminated by a slot with ``slot=0`` (typically written as ``{0}`` in C).
+There are no required slots.
+(So, the minimal input contains only the terminator; *slots* may not be
+``NULL``.)
+
+The *spec* argument is a duck-typed ModuleSpec-like object, meaning that any
+attributes defined for ``importlib.machinery.ModuleSpec`` have matching
+semantics.
+The ``name`` attribute is required, but this limitation may be lifted in the
+future.
+The ``name`` will be used *instead of* the ``Py_mod_name`` slot (just like
+``PyModule_FromDefAndSpec`` ignores ``PyModuleDef.m_name``).
+
+To simplify the implementation, the slots arrays for both
+``PyModule_FromSlotsAndSpec`` and the new export hook will only allow up to one
+``Py_mod_exec`` slot.
+(Arrays in ``PyModuleDef.m_slots`` may have more; this will not change.)
+
+For modules created without a ``PyModuleDef``, the ``Py_mod_create`` function
+will be called with ``NULL`` for the second argument.
+(In the future, if we find a use case for passing the input slots array, a new
+slot with an updated signature can be added.)
+
+Unlike the ``PyModExport_*`` hook, the *slots* array may be changed or
+destroyed after the ``PyModule_FromSlotsAndSpec`` call.
+(That is, Python must take a copy of all input data.)
+As an exception, any ``PyMethodDef`` array given by ``Py_mod_methods``
+must be statically allocated (or be otherwise guaranteed to outlive the
+objects created from it). This limitation may be lifted in the future.
+
+A new function will be added to run the ``exec`` slot(s) for a module -- like
+``PyModule_ExecDef``, but supporting modules created using slots,
+and not taking an explicit *def*:
+
+.. code-block:: c
+
+    int PyModule_Exec(PyObject *module)
+
+Calling this is required to fully initialize a module.
+``PyModule_FromSlotsAndSpec`` will *not* run it (just like
+``PyModule_FromDefAndSpec`` does not call ``PyModule_ExecDef``).
+
+For modules created from a *def*, calling this is equivalent to
+calling ``PyModule_ExecDef(module, PyModule_GetDef(module))``.
+
+
+Tokens
+------
+
+Module objects will optionally store a “token”: a ``void*`` pointer
+similar to ``Py_tp_token`` for types.
+
+If specified, using a new ``Py_mod_token`` slot, the module token must:
+
+- outlive the module, so it’s not reused for something else while the module
+  exists; and
+- “belong” to the extension module where the module lives, so it will not
+  clash with other extensions.
```

Comment:
```suggestion
- outlive the module, so it's not reused for something else while the module
  exists; and
- "belong" to the extension module where the module lives, so it will not
  clash with other extension modules.
```